File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        preMap = {i:[] for i in range(numCourses)}
        visitSet = set()

        for cur, pre in prerequisites:
            preMap[cur].append(pre)

        def dfs(cur):
            if cur in visitSet:
                return False
            if preMap[cur] == []:
                return True
            
            visitSet.add(cur)
            for pre in preMap[cur]:
                if not dfs(pre):
                    return False
            visitSet.remove(cur)
            preMap[cur] = []

            return True

        for cur in range(numCourses):
            if not dfs(cur):
                return False
        return True

File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(numCourses);
        for(auto v : prerequisites) {
            int a = v[0], b = v[1];
            graph[a].push_back(b);
        }
        vector<int> state(numCourses, 0); // 0 white, 1 gray, 2 black

        auto dfs = [&](auto &self, int u) {
            if(state[u] == 1) return false;
            if(state[u] == 2) return true;

            state[u] = 1;
            for(int v : graph[u]) {
                if(!self(self, v)) return false;
            }
            state[u] = 2;
            return true;
        };

        for(int i = 0; i < numCourses; i++) {
            if(!dfs(dfs, i)) return false; 
        }
        return true;
    }
};